### PR TITLE
Upgrade Go, Envoy and Envoy config versions

### PR DESCRIPTION
**Pull request by domgoodwin**
This PR:
- Upgrades Go to 1.18
- Upgrades Envoy container to 1.16
- Upgrades Envoy config to use v3 instead of the now deprecated v2
- Fixes the test runner:
  - kubebuilder download broken
  - Go version bump and go mod instead of go get
- Update the tests:
  - Envoy configmap test has now changed since we are on v3 config

### CODE REVIEW

File: .github/workflows/go.yml [hunk 1]
```diff
@@ -7,23 +7,19 @@ jobs:
...
       uses: actions/checkout@v1
 
     - name: Get dependencies
       run: |
-        go get -v -t -d ./...
-        if [ -f Gopkg.toml ]; then
-            curl https://raw.githubusercontent.com/golang/dep/master/install.sh | sh
-            dep ensure
-        fi
-        curl -sL https://go.kubebuilder.io/dl/2.2.0/linux/amd64 | tar -xz -C /tmp/
```
Comment (domgoodwin) [inline .github/workflows/go.yml:26]: This doesn't exist at this URL anymore

File: controllers/deployment.go [hunk 2]
```diff
@@ -108,7 +109,7 @@ func deployment(es *egressv1.ExternalService, configHash string) *appsv1.Deploym
 						{
 							Name: "gateway",
 							// TODO this version doesn't actually support UDP, we need 1.13 which isn't stable
```
Comment (milesbxf) [inline controllers/deployment.go:111]:
```suggestion
```

We can remove this comment

File: controllers/configmap.go [hunk 3]
```diff
@@ -6,24 +6,29 @@ import (
...
-	bootstrapv2 "github.com/envoyproxy/go-control-plane/envoy/config/bootstrap/v2"
-	accesslogfilterv2 "github.com/envoyproxy/go-control-plane/envoy/config/filter/accesslog/v2"
-	tcpproxyv2 "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/tcp_proxy/v2"
-	udpproxyv2 "github.com/envoyproxy/go-control-plane/envoy/config/filter/udp/udp_proxy/v2alpha"
+	accesslogfilterv3 "github.com/envoyproxy/go-control-plane/envoy/config/accesslog/v3"
+	bootstrap "github.com/envoyproxy/go-control-plane/envoy/config/bootstrap/v3"
+	envoyv3 "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
+	envoycorev3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	envoyendpoint "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
+	envoylistener "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
```
Comment (milesbxf) [inline controllers/configmap.go:15]: Bit of a nitpick and the original code doesn't do this well at all, but the convention we've generally used for aliasing for Envoy & K8s imports is to use the last word & version, e.g. core/v2 -> corev2


```suggestion
	corev2 "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
	accesslogv3 "github.com/envoyproxy/go-control-plane/envoy/config/accesslog/v3"
	bootstrapv3 "github.com/envoyproxy/go-control-plane/envoy/config/bootstrap/v3"
	clusterv3 "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
	corev3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
	endpointv3 "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
	listenerv3 "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
```

File: controllers/configmap.go [hunk 4]
```diff
@@ -83,85 +88,130 @@ const accessLogFormat = `[%START_TIME%] %BYTES_RECEIVED% %BYTES_SENT% %DURATION%
...
 			}},
 			AccessLogPath: "/dev/stdout",
 		},
-		StaticResources: &bootstrapv2.Bootstrap_StaticResources{},
+		StaticResources: &bootstrap.Bootstrap_StaticResources{},
 	}
 
 	for _, port := range es.Spec.Ports {
 		protocol := protocolToEnvoy(port.Protocol)
 		name := fmt.Sprintf("%s_%s_%s", es.Name, envoycorev2.SocketAddress_Protocol_name[int32(protocol)], strconv.Itoa(int(port.Port)))
```
Comment (milesbxf) [inline controllers/configmap.go:112]:
```suggestion
		name := fmt.Sprintf("%s_%s_%s", es.Name, envoycorev3.SocketAddress_Protocol_name[int32(protocol)], strconv.Itoa(int(port.Port)))
```

I think it'd make more sense to use corev3 here (assuming a similar mapping exists!)

**Review by milesbxf — APPROVED** [review]
Generally looks fine, a couple of small nitpicks

File: controllers/configmap.go (hunk 3 above)
Comment (domgoodwin) [reply controllers/configmap.go:15]: My worry here is below we import K8s resources as `corev1` which could get confusing between the two different packages?

Comment (milesbxf) [reply controllers/configmap.go:14]: Ooops, I didn't spot that - good point

**Review by milesbxf — APPROVED** [review]

**Pull request merged**